jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.cpp

```diff
@@ -743,7 +744,634 @@ ecma_builtin_string_prototype_object_match (ecma_value_t this_arg, /**< this arg
 
   return ret_value;
 } /* ecma_builtin_string_prototype_object_match */
-#endif /* !CONFIG_ECMA_COMPACT_PROFILE_DISABLE_REGEXP_BUILTIN */
+
+typedef struct
+{
+  /* General part. */
+  bool is_regexp; /**< whether we search a regexp or string */
+  bool is_global; /**< global search or not */
+  bool is_replace_callable; /**< replace part is callable or not */
+  ecma_value_t input_string; /**< input string */
+  ecma_length_t input_length; /**< input string length */
+  ecma_value_t regexp_or_search_string; /**< regular expression or search string
+                                         * depending on the value of is_regexp */
+  ecma_length_t match_start; /**< starting position of the match */
+  ecma_length_t match_end; /**< end position of the match */
+
+  /* Replace value callable part. */
+  ecma_object_t *replace_function_p;
+
+  /* Replace value string part. */
+  ecma_string_t *replace_string_p; /**< replace string */
+  lit_utf8_iterator_t replace_iterator; /**< replace string iterator */
+} replace_search_context;
+
+/**
+ * Generic helper function to append a substring at the end of a base string
+ *
+ * The base string can be kept or freed
+ *
+ * @return the constructed string
+ */
+static ecma_string_t *
+ecma_builtin_string_prototype_object_replace_append_substr (ecma_string_t *base_string_p, /**< base string */
+                                                            ecma_string_t *appended_string_p, /**< appended string */
+                                                            ecma_length_t start, /**< start position */
+                                                            ecma_length_t end, /**< end position */
+                                                            bool free_base_string) /**< free base string or not */
+{
+  ecma_string_t *appended_string;
```

Comment:
The variable name is very similar to argument's name `appended_string_p`. Maybe, `ret_string_p` would be better.